[SR-702][Parse] Disallow attributes before '#if'

rintaro opened the pull request:
#### What's in this pull request?

The parser should not allow any attributes before `#if`.

The following script used to be accepted:
```
struct Foo {
  public
#if true
  typealias Bar = Foo
#else
  typealias Bar = Foo
#endif
}
```
`public` were silently discarded in this case.

With this PR:
```
testif.swift:3:1: error: expected declaration
#if true
^
testif.swift:1:8: note: in declaration of 'Foo'
struct Foo {
       ^
```

#### Resolved bug number: ([SR-702](https://bugs.swift.org/browse/SR-702))

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

A smoke test on macOS does the following:

1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:

1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

rintaro commented:
Can somebody give me some early feedback?
Am I on the right way?

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
https://ci.swift.org/job/swift-PR-Linux-smoke-test/689/console
<details>
<summary>Linux failure looks unrelated</summary>
```
TestFoundation/TestNSNumber.swift:405:18: error: value of type 'NSNumber' has no member 'customPlaygroundQuickLook'
       let ql1 = NSNumber(value: 1234).customPlaygroundQuickLook
                 ^~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~
TestFoundation/TestNSNumber.swift:411:18: error: value of type 'NSNumber' has no member 'customPlaygroundQuickLook'
       let ql2 = NSNumber(value: Float(1.25)).customPlaygroundQuickLook
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~
TestFoundation/TestNSNumber.swift:417:18: error: value of type 'NSNumber' has no member 'customPlaygroundQuickLook'
       let ql3 = NSNumber(value: Double(1.25)).customPlaygroundQuickLook
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~
```
</details>

@jrose-apple Could you review this?

jrose-apple commented:
Doesn't this apply to `#sourceLocation` too?

jrose-apple commented:
Seems reasonable to me. We could probably improve the diagnostic too, but this fixes the issue at least.

rintaro commented:
> Doesn't this apply to #sourceLocation too?

Good point. I will handle that too. `#sourceLocation` and legacy `#line`

rintaro commented:
Updated to handle `#sourceLocation` and `#line`.
Also, added `TODO:` to improve diagnostic. We can implement that in follow up.

@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Test passed.
@jrose-apple Could you merge this if you are happy with this?

jrose-apple commented:
I have one outstanding question: how does this handle code completion in the pound-directives? The original code falls through to the end of the switch and does more work, but this just returns immediately. @benlangmuir, what's the correct thing to do here?

rintaro commented:
Actually, I don't know how the code completion things work. 😓 

As far as I understand from the existing code, as for `#if`, `#sourceLocation` and `#line`,
it seems there was no chance to enter these extra work branches after switch (related flags should never be true).
Maybe I'm wrong, not sure.

benlangmuir commented:
Sorry for the slow response!  Yeah this logic is a little hard to "see".  Could you make sure we have tests (or add them) for completing an override and an attribute inside an `#if`?

rintaro commented:
@benlangmuir Thank you for the comment!
Added test cases for completion combined with `#if`

How about directives itself? It seems, even in master branch, the completion doesn't work for:
```
// expecting "sourceLocation" and "if"
##^TOKEN^#

// expecting "file: "
#sourceLocation(#^TOKEN^#)

// expecting  "os(...)", "arch(...)", or such.
#if #^TOKEN^#
```

rintaro commented on test/IDE/complete_pound_decl.swift:
```diff
@@ -0,0 +1,53 @@
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=POUND_ELSE_MEMATTR | %FileCheck %s -check-prefix=ATTR
+
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=POUND_IF_GBLATTR | %FileCheck %s -check-prefix=ATTR
+// FIXME: RUN: not --crash %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=POUND_IF_GBLNAME 
```
This does crash with:
<details>
<summary>output</summary>
```
found code completion token POUND_IF_GBLNAME at offset 1647
found code completion token POUND_IF_GBLNAME at offset 1647
invalid IfConfigStmt end location
(#if_stmt
  (#if:
    (boolean_literal_expr type='<null>' value=true)
    (elements
      (func_decl "bar()" type='<null type>'
        (parameter_list))

      (top_level_code_decl
        (brace_stmt))
))0  swift-ide-test  0x0000000002f49098 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift-ide-test  0x0000000002f47896 llvm::sys::RunSignalHandlers() + 54
2  swift-ide-test  0x0000000002f49bc6
3  libpthread.so.0 0x00007fe94abe1d10
4  libc.so.6       0x00007fe9493a91c7 gsignal + 55
5  libc.so.6       0x00007fe9493aae2a abort + 362
6  swift-ide-test  0x0000000000bbba03
7  swift-ide-test  0x0000000000bd0efb
8  swift-ide-test  0x0000000000bcf6f6
9  swift-ide-test  0x0000000000bcf2f4 swift::Decl::walk(swift::ASTWalker&) + 20
10 swift-ide-test  0x0000000000c6a07e swift::SourceFile::walk(swift::ASTWalker&) + 174
11 swift-ide-test  0x0000000000bb1fc4 swift::verify(swift::SourceFile&) + 52
12 swift-ide-test  0x00000000008ca0c3 swift::Parser::parseTopLevel() + 563
13 swift-ide-test  0x00000000008ff130 swift::parseIntoSourceFile(swift::SourceFile&, unsigned int, bool*, swift::SILParserState*, swift::PersistentParserState*, swift::DelayedParsingCallbacks*) + 208
14 swift-ide-test  0x00000000007aa646 swift::CompilerInstance::performSema() + 3254
15 swift-ide-test  0x000000000074d2d0 main + 34704
16 libc.so.6       0x00007fe949394ac0 __libc_start_main + 240
17 swift-ide-test  0x0000000000741639 _start + 41
Stack dump:
0.	Program arguments: /home/rintaro/Documents/swift-oss/build/Ninja-ReleaseAssert/swift-linux-x86_64/bin/swift-ide-test -code-completion -source-filename /home/rintaro/Documents/swift-oss/swift/test/IDE/complete_pound_decl.swift -code-completion-token=POUND_IF_GBLNAME 
1.	With parser at source location: /home/rintaro/Documents/swift-oss/swift/test/IDE/complete_pound_decl.swift:54:1
2.	While walking into decl declaration 0x5fde500 at /home/rintaro/Documents/swift-oss/swift/test/IDE/complete_pound_decl.swift:43:1
Aborted (core dumped)
```
</details>

benlangmuir commented:
@rintaro right, we only have completion for a few specific things after #.  It wouldn't be very hard to add more, but that's a separate issue.  Nice find with the crash, could you file a bug?

rintaro commented:
@benlangmuir Filed as [SR-2364](https://bugs.swift.org/browse/SR-2364)

jrose-apple commented:
Is this ready to land, @benlangmuir?

benlangmuir commented:
LGTM, thanks!

benlangmuir commented:
@swift-ci Please smoke test
